File: LeetCode/Easy/0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.py
class Solution(object):
    def maxProfit(self, prices):
        # d[i] 는 i번째 까지 최대 이익 팔거나 or 팔지 않거나
        d = [0] * len(prices)
        d[0] = 0
        # d[1] = prices[1] - prices[0]
        min_price = prices[0]
        for i in range(1, len(prices)):
            d[i] = max(d[i - 1], prices[i] - min_price)
            if min_price > prices[i]:
                min_price = prices[i]

        return d[len(prices) - 1]

File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        if (prices.size() < 2) return 0;
        int ans = 0;
        int l = 0;
        int r = 1;
        int n = prices.size();
        while (r < n){
            if (prices[l] >= prices[r]){
                l=r; r++;
            } else {
                ans = max(ans,prices[r] - prices[l]);
                r++;
            }
        }
        return ans;
    }
};